File: 0337_House Robber III/HouseRobberIII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        if(root == null) return 0;
        int val = root.val;
        int ll = root.left != null ? rob(root.left.left) : 0;
        int lr = root.left != null ? rob(root.left.right) : 0;
        int rl = root.right != null ? rob(root.right.left) : 0;
        int rr = root.right != null ? rob(root.right.right) : 0;
        return Math.max(val + ll + lr + rl + rr, rob(root.left) + rob(root.right));
    }
}

File: 0337_House Robber III/HouseRobberIII.py
class Solution:
  def rob(self, root):
    def rob(root):
      if not root: return 0, 0, 0
      l, ll, lr = rob(root.left)
      r, rl, rr = rob(root.right)
      return max(root.val + ll + lr + rl + rr, l + r), l, r
    
    return rob(root)[0]